Plugins can now define custom validation rules

mohsin opened the pull request:
Added a new feature by which plugins can now define custom validators. For example, the Responsiv.Currency plugin can now declare a new validation for currency by adding a code block like:
```php
/**
 * Register any new validators
 * @return array
 */
public function registerValidator()
{
    return [
        'currency' => function ($attribute, $value, $parameter, $validator) {
            // return true if $value is valid currency format
        }
    ];
}
```

mohsin committed:
Plugins can now define custom validators

LukeTowers commented on modules/system/ServiceProvider.php:
```diff
@@ -575,6 +576,15 @@ protected function registerValidator()
                 return strtr($message, [':values' => implode(', ', $parameters)]);
             });
+
+            $plugins = PluginManager::instance()->getRegistrationMethodValues('registerValidator');
```
```suggestion
            $plugins = PluginManager::instance()->getRegistrationMethodValues('registerValidators');
```
Also needs PR to docs repo

mohsin replied:
Added PR to docs repo to support this PR.

mohsin committed:
Update modules/system/ServiceProvider.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo commented:
@SaifurRahmanMohsin @LukeTowers should this not perhaps be `registerValidationRules`, as it's a rule being registered, not an entire Validator instance?

mohsin commented:
That sounds better to me. Shall I make commits to amend it? (I've checked the allow maintainers to edit thingy so you can go ahead and do it as well).

bennothommo commented:
@SaifurRahmanMohsin feel free to make the change if @LukeTowers agrees with the wording.

One thing that might be handy is to support [Rule objects](https://github.com/octobercms/library/pull/455) being implemented as well, like so:

```php
/**
 * Register any new validators
 * @return array
 */
public function registerValidator()
{
    return [
        'currency' => CurrencyRule::class,
    ];
}
```

This way, people can use self-contained rule objects that have their own validation messages defined in the class as well.

LukeTowers commented:
`registerValidationRules` works for me, and supporting Rule classes is a good idea

mohsin committed:
Renamed to registerValidationRules

Renamed to a more appropriate name based on the suggestions by @bennothommo and @LukeTowers.

mohsin committed:
Support for Rule objects

Plugins can now register validation rules using Rule objects as well.

mohsin commented:
Changes have been made @bennothommo @LukeTowers

LukeTowers commented:
@SaifurRahmanMohsin would you be willing to add some unit tests for the supported ways of defining custom validators (rule object vs callable)?

mohsin commented:
@LukeTowers I would like to but I'm unable to get the tests suite working on my machine for some reason. I just do this:
```
git clone <this-repo> // Checks out develop branch
composer up
composer test
```

This fails and throws this:
```
1) BelongsToManyModelTest::testSetRelationValue
Illuminate\Database\QueryException: SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: database_tester_authors_roles.author_id, database_tester_authors_roles.role_id (SQL: insert into "database_tester_authors_roles" ("author_id", "role_id") values (1, 3))
```

So I checked git status and found some of the files were changes under modules due to composer up. So I tried a:
```
git checkout modules/
composer test
```

Now I get a single error:
```
1) FilterTest::testRestrictedScopeWithUserWithNoPermissions
ErrorException: Class 'October\Rain\Support\Facades\Mail' not found

/Users/Mohsin/Desktop/october/vendor/laravel/framework/src/Illuminate/Foundation/AliasLoader.php:80
/Users/Mohsin/Desktop/october/tests/PluginTestCase.php:104

ERRORS!
Tests: 24, Assertions: 96, Errors: 1.
Script phpunit --stop-on-failure handling the test event returned with error code 2
```

I understand tests run on their own environment so I wouldn't need to run `php artisan october:install` correct? (I tried running tests with that too but it fails again).

So I'm just unable to even get the tests working initially so writing a new test becomes harder. I tried coding it using --filter but then the fixtures were not setting up so I'm fairly certain that I need to get the entire test suite working for it to work?

LukeTowers commented:
@SaifurRahmanMohsin you need the develop version of october/rain, do `dev-develop as 1.1` for october rain, run composer update, then recheckout the develop branch for the october repo

mohsin commented:
@LukeTowers Cool, that worked with just one test failure. Thanks! I really think this should to be documented [somewhere](https://github.com/octobercms/october/blob/develop/.github/CONTRIBUTING.md). Alright, I'll try to get this test made.

mohsin committed:
Added unit tests for custom validation rules in plugins

LukeTowers commented on tests/fixtures/plugins/october/tester/Plugin.php:
```diff
@@ -61,4 +61,14 @@ public function registerNavigation()
+    {
+        return [
+            'uppercase' => function ($attribute, $value, $parameters, $validator) {
+                return strtoupper($value) == $value;
```
```suggestion
                return strtoupper($value) === $value;
```
Any reason why you used the weak comparison operator instead of the strict one or was it just habit?

mohsin replied:
Just a habit @LukeTowers

LukeTowers replied:
@SaifurRahmanMohsin in general I would recommend that you reverse that habit, `===` is going to save you from a lot of bugs that `==` can accidentally lead to.

LukeTowers commented on tests/unit/plugins/system/CustomValidatorRulesTest.php:
```diff
@@ -0,0 +1,57 @@
+        parent::setUp();
+
+        include_once base_path() . '/tests/fixtures/plugins/october/tester/rules/BeLikeBobRule.php';
+        include_once base_path().'/tests/fixtures/plugins/october/tester/Plugin.php';
```
```suggestion
        include_once base_path() . '/tests/fixtures/plugins/october/tester/Plugin.php';
```

Commit:
Update tests/unit/plugins/system/CustomValidatorRulesTest.php

mohsin committed:
Update tests/fixtures/plugins/october/tester/Plugin.php

Co-authored-by: Luke Towers <github@luketowers.ca>

LukeTowers commented on tests/fixtures/plugins/october/tester/Plugin.php:
```diff
@@ -61,4 +61,14 @@ public function registerNavigation()
+            'uppercase' => function ($attribute, $value, $parameters, $validator) {
+                return strtoupper($value) === $value;
+            },
+            'be_like_bob' => \October\Tester\Rules\BeLikeBobRule::class
```
```suggestion
            'be_like_bob' => \October\Tester\Rules\BeLikeBobRule::class,
```

LukeTowers commented on tests/fixtures/plugins/october/tester/rules/BeLikeBobRule.php:
```diff
@@ -0,0 +1,41 @@
+     */
+    public function passes($attribute, $value)
+    {
+        return $value == 'bob';
```
```suggestion
        return $value === 'bob';
```

Commit:
Update tests/fixtures/plugins/october/tester/Plugin.php

Commit:
Update tests/fixtures/plugins/october/tester/rules/BeLikeBobRule.php

LukeTowers commented:
@SaifurRahmanMohsin is there an associated docs PR that you can link me to?

mohsin commented:
Hey Luke, that PR [is here](https://github.com/octobercms/docs/pull/513).

LukeTowers commented:
@SaifurRahmanMohsin thanks, I've added some comments to it

daftspunk commented:
Hey @SaifurRahmanMohsin , I'm not sure this adds much beyond what is described in the documentation on extending validation rules. Plugins can already do this via the boot method to interface with Laravel's validator.

```
    public function boot()
    {
        Validator::extend('foo', function($attribute, $value, $parameters) {
            return $value == 'foo';
        });
    }
```

Let me know if I am missing something here and we can revisit it. Thanks 🙏
